python/cudf/cudf/utils/ioutils.py

```diff
@@ -1065,3 +1081,108 @@ def buffer_write_lines(buf, lines):
     if any(isinstance(x, str) for x in lines):
         lines = [str(x) for x in lines]
     buf.write("\n".join(lines))
+
+
+def _check_contains_null(val):
+    if isinstance(val, bytes):
+        for byte in val:
+            if isinstance(byte, bytes):
+                compare_to = chr(0)
+            else:
+                compare_to = 0
+            if byte == compare_to:
+                return True
+    elif isinstance(val, str):
+        return "\x00" in val
+    return False
+
+
+def _check_filters(filters, check_null_strings=True):
+    """
+    Check if filters are well-formed.
+    """
+    if filters is not None:
+        if len(filters) == 0 or any(len(f) == 0 for f in filters):
+            raise ValueError("Malformed filters")
+        if isinstance(filters[0][0], str):
+            # We have encountered the situation where we have one nesting level
+            # too few:
+            #   We have [(,,), ..] instead of [[(,,), ..]]
+            filters = [filters]
+        if check_null_strings:
+            for conjunction in filters:
+                for col, op, val in conjunction:
+                    if (
+                        isinstance(val, list)
+                        and all(_check_contains_null(v) for v in val)
+                        or _check_contains_null(val)
+                    ):
+                        raise NotImplementedError(
+                            "Null-terminated binary strings are not supported "
+                            "as filter values."
+                        )
+    return filters
+
+
+def _convert_single_predicate(col, op, val):
+    field = ds.field(col)
+
+    if op == "=" or op == "==":
+        return field == val
+    elif op == "!=":
+        return field != val
+    elif op == "<":
+        return field < val
+    elif op == ">":
+        return field > val
+    elif op == "<=":
+        return field <= val
+    elif op == ">=":
+        return field >= val
+    elif op == "in":
+        return field.isin(val)
+    elif op == "not in":
+        return ~field.isin(val)
+    else:
+        raise ValueError(
+            '"{0}" is not a valid operator in predicates.'.format(
+                (col, op, val)
+            )
+        )
+
+
+def _convert_single_conjunction(conjunction):
+    and_exprs = [
+        _convert_single_predicate(col, op, val) for col, op, val in conjunction
+    ]
+
+    expr = and_exprs[0]
+    if len(and_exprs) > 1:
+        for and_expr in and_exprs[1:]:
+            expr = ds.AndExpression(expr, and_expr)
+    return expr
+
+
+# TODO: Use PyArrow once filters-to-expression conversion is added to
+# API by resolving https://issues.apache.org/jira/browse/ARROW-9672
```

Comment:
Personally, I think it is fine to just use the private `_filters_to_expression` for now, instead of copying its source here (100+ lines). Eg also dask is using it. We should of course still discuss making it public in ARROW-9672

Yes, it's a private utility (but we can guarantee we won't simply remove it in the next version), but on the other hand, now you are using AndExpression/OrExpression APIs, which are also subject to change as Wes noted on the JIRA (https://issues.apache.org/jira/browse/ARROW-9672).

So I would personally feel safer that cudf used `_filters_to_expression` and then pyarrow can ensure this method continues to return *some* expression object that can be used in the dataset API, than that pyarrow needs to guarantee that the exact Expression API stays the same.